springboot/src/main/java/io/temporal/samples/springboot/update/PurchaseWorkflowImpl.java

```diff
@@ -0,0 +1,64 @@
+/*
+ *  Copyright (c) 2020 Temporal Technologies, Inc. All Rights Reserved
+ *
+ *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ *  Modifications copyright (C) 2017 Uber Technologies, Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
+ *  use this file except in compliance with the License. A copy of the License is
+ *  located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ *  or in the "license" file accompanying this file. This file is distributed on
+ *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *  express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package io.temporal.samples.springboot.update;
+
+import io.temporal.activity.LocalActivityOptions;
+import io.temporal.samples.springboot.update.model.Purchase;
+import io.temporal.spring.boot.WorkflowImpl;
+import io.temporal.workflow.Workflow;
+import java.time.Duration;
+
+@WorkflowImpl(taskQueues = "UpdateSampleTaskQueue")
+public class PurchaseWorkflowImpl implements PurchaseWorkflow {
+
+  private boolean newPurchase = false;
+  private boolean exit = false;
+  private PurchaseActivities activities =
+      Workflow.newLocalActivityStub(
+          PurchaseActivities.class,
+          LocalActivityOptions.newBuilder().setStartToCloseTimeout(Duration.ofSeconds(2)).build());
+
+  @Override
+  public void start() {
+    // for sake of sample we only wait for a single purchase or exit signal
+    Workflow.await(() -> newPurchase || exit);
+  }
+
+  @Override
+  public boolean makePurchase(Purchase purchase) {
+    return activities.makePurchase(purchase);
+  }
+
+  @Override
+  public void makePurchaseValidator(Purchase purchase) {
+    if (!activities.isProductInStockForPurchase(purchase)) {
```

Comment:
Local activities cannot be run in validators

Reply:
they have the same restriction as queries

Reply:
it works as in it gets called ok. didnt see failure

Reply:
it is not supported and It will not replay correctly

Reply:
if that is enforced tho idk why? would be very restrictive if could not check db for example to validate update

Reply:
here is event history that shows that local activity called in validator does get executed if that helps.
this is the path when you enter in amount thats greater than that in inventory:

<img width="1473" alt="Screen Shot 2023-06-27 at 1 36 10 AM" src="https://github.com/temporalio/samples-java/assets/119422/d4652d58-389a-42c1-900b-bd57b26bd873">

Reply:
ill add WorkflowReplayer test for this tomorrow and see 

but if it does not work i think the validators are pretty weak honestly. for use cases like this one not being able to check db or invoke some api to validate is imho useless

Reply:
validators cannot do anything that modifies workflow state like executing an activity or modifying a workflow variable.

They are only able to read workflow state and update inputs. There is desire to support running local activities in update validators, but the server does not support this yet.

Reply:
ok will move that check to update method then. so if it does nit update wf state its ok like in this example?

Reply:
this example runs a local activity right?  A local activity modifies workflow state so it is not allowed. The rules for what is allowed in a validator are the same as for a query handler.

Reply:
@Quinn-With-Two-Ns updated the code to not invoke LA from update validator. let me know if ok now

Reply:
just wanted to add, having to throw TemporalFailure from update handler is weird. If this is done in workflow code the wf execution would fail, same if you did in signal handlers, so thats why it just feels weird that you have to do it and it does not fail execution just the update

Reply:
We'll likely not support local activities in validators but there has been talk of adding a different concept to SDKs which is backed by activities but doesn't generate history events that can be used in both query handlers and update validators.

Reply:
We should throw in Java (and Go) if any commands are being generated in query handlers and update validators (and side-effects FTR).

Reply:
>ok. should we then open issue for it in java sdk to throw error and not allow temporal apis?
Yes, [we already do ](https://github.com/temporalio/sdk-java/issues/1748)

>  just wanted to add, having to throw TemporalFailure from update handler is weird
Do you have a suggestion on how you would implement it differently? I think the issue is your thinking of the update handler as part of the workflow like signal, but really its another sub request running inside a workflow that can succeed/fail on its own.

Reply:
I think if its possible to have a specific error idk like RejectUpdateException or something like that would be easier for users to understand (and if you throw TemporalFailure for it to still fail workflow if user really wants to do that for some reason).

Reply:
So that doesn't really work because if you ran an activity in a update handle and it failed by default if would fail your whole workflow instead of just your update request.